### PR TITLE
eABF implementation

### PR DESCRIPTION
**Pull request by HanatoK**
##### Description

This pull request implements the extended adaptive biasing force(eABF)/dynamic reference restraining(DRR)  method.

##### Target release

I would like my code to appear in release v2.4

##### Type of contribution

- [ ] changes to code or doc authored by PLUMED developers
- [ ] changes to a module not authored by you
- [X] new module contribution or edit of a module authored by you

##### Copyright

- [X] I agree to transfer the copyright of the code I have written to the PLUMED developers or to the author of the code I am modifying.

##### Tests

Make sure these boxes are checked before submitting your pull request:

- [X] I added a new regtest or modified an existing regtest to validate my changes.
- [ ] I verified that all regtests are passed successfully on Travis-CI.

The Travis-CI test is failed due to following:
[drr/colvar_UIestimator.h:854] (error) :plmd_missing_guard: missing ifndef guard
[drr/DRR.h:335] (error) :plmd_missing_guard: missing ifndef guard

I added ifndef guard but still failed. I can't figure out why.

### CODE REVIEW

File: src/drr/DRR.h [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ */
+// Build requirement: boost, c++11 compatible compiler.
+#ifdef __PLUMED_HAS_BOOST_SERIALIZATION
```
Comment (GiovanniBussi) [inline src/drr/DRR.h]: @HanatoK I think plumedcheck wants the guard to be the first preprocessor command. Anyway, you can generate guards using a script (more later in the general comments)

File: src/drr/drrtool.cpp [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+  void extractdrr(const std::vector<std::string> &filename);
+  void mergewindows(const std::vector<std::string> &filename);
+  std::string description() const { return "Extract or merge the drrstate files."; }
+
+private:
+  bool verbosity;
+  const std::string suffix{".drrstate"};
+};
+
+PLUMED_REGISTER_CLTOOL(drrtool, "drrtool")
```
Comment (GiovanniBussi) [inline src/drr/drrtool.cpp]: We are in the process of deciding rules for namespaces. Perhaps you should change this to
````
PLUMED_REGISTER_CLTOOL(drrtool, "drr_tool")
````
(notice  the underscore)

**Comment by GiovanniBussi** [comment]
@HanatoK thanks a lot for your contribution!

You only added an extra module, so there are very few things to fix.

Please rebase on branch `v2.4`, otherwise your contribution will end up in `v2.5` (current master)

For contributed modules, we did not write guidelines yet but we will likely enforce some rule. You might be asked later to write a short markdown page explaining what the module does (something like [this one](https://plumed.github.io/doc-isdb/user-doc/html/_i_s_d_b.html).

Thanks again!

Giovanni

**Comment by GiovanniBussi** [comment]
@HanatoK I forgot this:
Please add a COPYRIGHT file to src/drr and rerun `./headers.sh` from `src`. This should fix your guards and also the copyright statements at the beginning of each file. See for instance `src/lepton`.

Giovanni

File: regtest/drr/rt-eabf/config [hunk 3]
```diff
@@ -0,0 +1,2 @@
+type=driver
+arg="--plumed plumed.dat --trajectory-stride 1 --timestep 0.005 --ixyz ala12_trajectory.xyz --dump-forces forces --dump-forces-fmt=%10.6f"
```
Comment (GiovanniBussi) [inline regtest/drr/rt-eabf/config]: @HanatoK you should add this line to all the `config` files of your regtests:
````
plumed_modules=drr
````

This is to tell plumed that this regtest should be skipped with module drr is not enabled

**Comment by HanatoK** [comment]
@GiovanniBussi When I ran header.sh, other files in src/pamm are also changed. Is this a problem?
git tells me I can't rebase to origin/v2.4 because of those changes.

**Comment by GiovanniBussi** [comment]
Sorry my fault. You should have done
````
cd src
./headers.sh ddr
````
Or you can remove the changes to `src/pamm` from the commit (with `git commit --amend`)

**Comment by HanatoK** [comment]
@GiovanniBussi 
I didn't push src/pamm changes and just checkout them from remote branch.
It seems I can't run `git rebase origin/v2.4` directly. There are many conflicts. Does I need to create a new fork of v2.4 and put my changes there?

**Comment by GiovanniBussi** [comment]
You have to rebase only your changes (not those that are in master). You can use `rebase --onto`. I typically find it easier to do something like
````
git checkout v2.4
git merge origin/v2.4
git cherry-pick xxx
git cherry-pick yyy
etc ...
````
Where xxx and yyy are the hash of your commits (older to newer). I don't expect any conflict since you added an external module.

**Comment by GiovanniBussi** [comment]
Moved to #276

**Pull request closed without merging**